atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/expectations/KeyValueServiceDataTracker.java

```diff
@@ -0,0 +1,132 @@
+                                .maximumBytesKvsCallInfo(
+                                        Optional.ofNullable(maximumBytesKvsCallInfoByTable.get(tableRef)))
+                                .build()));
+    }
+
+    /**
+     * Tracks an effectively completed kvs read method call for some {@link TableReference}.
+     * Effectively completed refers to an eager call (i.e. does not spawn futures or iterators for later consumption).
+     */
+    public void recordReadForTable(TableReference tableRef, String methodName, long bytesRead) {
+        updateKvsMethodTalliesByTable(tableRef, bytesRead, methodName);
+    }
+
+    /**
+     * Track an effectively completed kvs read method call with no {@link TableReference} information.
+     * Effectively completed refers to an eager call (i.e. does not spawn futures or iterators for later consumption).
+     */
+    public void recordTableAgnosticRead(String methodName, long bytesRead) {
+        updateKvsMethodTalliesOverall(bytesRead, methodName);
+    }
+
+    /**
+     * Registers that a kvs read method was called for some {@link TableReference} and provides a consumer that will
+     * register bytes read as they come (e.g. use case is for calls returning iterators).
+     */
+    public Consumer<Long> recordCallForTable(TableReference tableRef) {
+        kvsCallsOverall.increment();
+        kvsCallByTable.incrementAndGet(tableRef);
+        return bytesRead -> {
+            bytesReadOverall.add(bytesRead);
+            bytesReadByTable.addAndGet(tableRef, bytesRead);
+        };
+    }
+
+    private void updateKvsMethodTalliesOverall(long bytesRead, String methodName) {
+        kvsCallsOverall.increment();
+        bytesReadOverall.add(bytesRead);
+        maximumBytesKvsCallInfoOverall.updateAndGet(currentMaybeCall -> Comparators.max(
+                currentMaybeCall,
+                Optional.of(ImmutableKvsCallReadInfo.builder()
+                        .bytesRead(bytesRead)
+                        .methodName(methodName)
+                        .build()),
+                Comparators.emptiesFirst(Comparator.naturalOrder())));
+    }
+
+    private void updateKvsMethodTalliesByTable(TableReference tableRef, long bytesRead, String methodName) {
+        KvsCallReadInfo callInfo = ImmutableKvsCallReadInfo.builder()
+                .bytesRead(bytesRead)
+                .methodName(methodName)
+                .build();
+
+        kvsCallsOverall.increment();
+        bytesReadOverall.add(bytesRead);
+        maximumBytesKvsCallInfoOverall.updateAndGet(currentMaybeCall -> Comparators.max(
+                currentMaybeCall, Optional.of(callInfo), Comparators.emptiesFirst(Comparator.naturalOrder())));
+
+        kvsCallByTable.incrementAndGet(tableRef);
+        bytesReadByTable.addAndGet(tableRef, callInfo.bytesRead());
+        maximumBytesKvsCallInfoByTable.merge(tableRef, callInfo, Comparators::max);
+    }
+}
```

Comment:
There is some code duplication here now, but I don't think it's that bad.

I wanted to only build `KvsReadCallInfo` once (and in this class not from the tracking kvs, so as not to clutter the latter).